### ISSUE TITLE
Inputs are being fired twice

### ISSUE THREAD

**Issue by ThowV** [Unity-Technologies/InputSystem#959]
If you follow the documentation on the https://docs.unity3d.com/Packages/com.unity.inputsystem@1.0/manual/QuickStartGuide.html site, you'll get the following quirky behaviour: 
Using any input that has the Press Only Interaction set on it will call the mapped event twice. This happens to me every time I create a new project in Unity 2020.1.0a13.

Hope this can get resolved, thanks in advance.

**Comment by ThowV** [comment]
I fixed the issue by using one of the sample projects you can download from the package manager.

**Comment by diegoecamacho** [comment]
This happened to me as well, I'm running Unity 2019.3.0b12 and Input System package 1.00 preview.3

I found i had this issue when i used the "Player Input" component.

I got around this by doing it through script. i created an object from my input action asset c# class 
and handled all event registration through script.

Hope this helps, doing great work guys!

**Comment by Sarion56** [comment]
I am also getting this issue. Removing my Player Input component didn't help.

**Comment by diegoecamacho** [comment]
This is how I got around this issue, while it gets resolved.

---

Empty class just to be able to set the events in editor

<pre><code>
[System.Serializable]
/// Remove the '' from this line
public class InputSystemEvent : UnityEvent'<'InputAction.CallbackContext>'>' {}
</code></pre>

The replacement class for Player Input.
<pre><code>
public class InputManager : MonoBehaviour   
{   
    ///you need to generate a c# class from your input action asset
    public MyInputActionAsset inputAsset { get; private set;}
    
    public InputSystemEvent MovementEvent;

    void OnEnable() {    
        inputAsset  = new MyInputActionAsset();    
        inputAsset.Enable();
        
        /// register your unity event with the action you create 
        inputAsset.ActionMap.Movement.performed += MovementEvent.Invoke;
    }
    
    void OnDisable() {

      ///Unregister unity event to be safe
       MyInputActionAsset.ActionMap.Movement.performed -= MovementEvent.Invoke;
       MyInputActionAsset.Disable();
   }
}
</code></pre>

**Comment by OmegaRogue** [comment]
> This is how I got around this issue, while it gets resolved.
> 
> Empty class just to be able to set the events in editor
> 
> ```
> 
> [System.Serializable]
> /// Remove the '' from this line
> public class InputSystemEvent : UnityEvent'<'InputAction.CallbackContext>'>' {}
> ```
> 
> The replacement class for Player Input.
> 
> ```
> 
> public class InputManager : MonoBehaviour   
> {   
>     ///you need to generate a c# class from your input action asset
>     public MyInputActionAsset inputAsset { get; private set;}
>     
>     public InputSystemEvent MovementEvent;
> 
>     void OnEnable() {    
>         inputAsset  = new MyInputActionAsset();    
>         inputAsset.Enable();
>         
>         /// register your unity event with the action you create 
>         inputAsset.ActionMap.Movement.performed += MovementEvent.Invoke;
>     }
>     
>     void OnDisable() {
> 
>       ///Unregister unity event to be safe
>        MyInputActionAsset.ActionMap.Movement.performed -= MovementEvent.Invoke;
>        MyInputActionAsset.Disable();
>    }
> }
> ```

what is `MyInputActionAsset`? its showing up as red for me

**Comment by OmegaRogue** [comment]
I think I found a solution working with PlayerInput directly.
I looked through the Code of PlayerInput and found the code for handling Input when set to UnityEvents:
```cs
case PlayerNotifications.InvokeUnityEvents:
                {
                    // Hook up all action events.
                    if (m_ActionEvents != null)
                    {
                        foreach (var actionEvent in m_ActionEvents)
                        {
                            var id = actionEvent.actionId;
                            if (string.IsNullOrEmpty(id))
                                continue;

                            // Find action for event.
                            var action = m_Actions.FindAction(id);
                            if (action != null)
                            {
                                ////REVIEW: really wish we had a single callback
                                action.performed += actionEvent.Invoke;
                                action.canceled += actionEvent.Invoke;
                                action.started += actionEvent.Invoke;
                            }
                            else
                            {
                                // Cannot find action. Log error.
                                if (!string.IsNullOrEmpty(actionEvent.actionName))
                                {
                                    // We have an action name. Show in message.
                                    Debug.LogError(
                                        $"Cannot find action '{actionEvent.actionName}' with ID '{actionEvent.actionId}' in '{actions}",
                                        this);
                                }
                                else
                                {
                                    // We have no action name. Best we have is ID.
                                    Debug.LogError(
                                        $"Cannot find action with ID '{actionEvent.actionId}' in '{actions}",
                                        this);
                                }
                            }
                        }
                    }
                    break;
                }
```
This part is the important part:
```cs
////REVIEW: really wish we had a single callback
                                action.performed += actionEvent.Invoke;
                                action.canceled += actionEvent.Invoke;
                                action.started += actionEvent.Invoke;
```
The reason it triggers multiple times is that the Method is Hooked multiple times.
So then i looked if there was anything I could use to check which of these hooks called the method and found these:
```cs
 public void TestInput(InputAction.CallbackContext ctx)
    {
        if(ctx.canceled);
        if(ctx.performed);
        if(ctx.started);
        if(ctx.phase == InputActionPhase.Canceled);
        if(ctx.phase == InputActionPhase.Disabled);
        if(ctx.phase == InputActionPhase.Performed);
        if(ctx.phase == InputActionPhase.Started);
        if(ctx.phase == InputActionPhase.Waiting);
    }
```
so to fix the problem, you can for example do this:
```cs
public void TestInput(InputAction.CallbackContext ctx)
    {
        if(!ctx.performed)
            return;
	}
```
or this:
```cs
public void TestInput(InputAction.CallbackContext ctx)
    {
        if(ctx.phase != InputActionPhase.Performed)
            return;
	}
```

**Comment by fabienjuif** [comment]
Hey @OmegaRogue 👋 

I am doing quite the same thing but with a different phase : `if (!context.started) return;`
Why do you think "performed" is better?

**Comment by OmegaRogue** [comment]
> Hey @OmegaRogue 👋
> 
> I am doing quite the same thing but with a different phase : `if (!context.started) return;`
> Why do you think "performed" is better?

I don't know which is better, i used `ctx.performed` because @diegoecamacho hooked `Performed` in their Workaround

**Comment by OmegaRogue** [comment]
The order they are called in is:
```cs
InputActionPhase.Started
InputActionPhase.Performed
InputActionPhase.Canceled
```

**Comment by OmegaRogue** [comment]
I looked at the code of `InputAction` and `phase` has these xml docs:
```cs
/// <summary>
/// The current phase of the action.
/// </summary>
/// <remarks>
/// When listening for control input and when responding to control value changes,
/// actions will go through several possible phases.
///
/// In general, when an action starts receiving input, it will go to <see cref="InputActionPhase.Started"/>
/// and when it stops receiving input, it will go to <see cref="InputActionPhase.Canceled"/>.
/// When <see cref="InputActionPhase.Performed"/> is used depends primarily on the type
/// of action. <see cref="InputActionType.Value"/> will trigger <see cref="InputActionPhase.Performed"/>
/// whenever the value of the control changes (including the first time; i.e. it will first
/// trigger <see cref="InputActionPhase.Started"/> and then <see cref="InputActionPhase.Performed"/>
/// right after) whereas <see cref="InputActionType.Button"/> will trigger <see cref="InputActionPhase.Performed"/>
/// as soon as the button press threshold (<see cref="InputSettings.buttonPressThreshold"/>)
/// has been crossed.
///
/// Note that both interactions and the action <see cref="type"/> can affect the phases
/// that an action goes through. <see cref="InputActionType.PassThrough"/> actions will
/// only ever use <see cref="InputActionPhase.Performed"/> and not go to <see
/// cref="InputActionPhase.Started"/> or <see cref="InputActionPhase.Canceled"/> (as
/// pass-through actions do not follow the start-performed-canceled model in general).
/// Also, interactions can choose their
///
/// While an action is disabled, its phase is <see cref="InputActionPhase.Disabled"/>.
/// </remarks>
public InputActionPhase phase => currentState.phase;
```
that should help choosing which one to check for.
The code is in
`Library/PackageCache/com.unity.inputsystem@1.0.0-preview.3/InputSystem/Actions/InputAction.cs`
and PlayerInput is in
`Library/PackageCache/com.unity.inputsystem@1.0.0-preview.3/InputSystem/Plugins/PlayerInput/PlayerInput.cs`

**Comment by stefanunity** [comment]
Hello,

we're going to close all the github issues and will concentrate bug information in the Unity public issue tracker https://issuetracker.unity3d.com/product/unity/issues?project=116&status=1&unity_version=&view=hottest.

If you are still experienceing this issue in the latest package version please do file a bug report as described here https://unity3d.com/unity/qa/bug-reporting
For general feedback and discussions you can go to https://forum.unity.com/forums/new-input-system.103/

Thank you!

**Comment by ps2goat** [comment]
I was trying to get the buttons to work, too. I have a `bool` value I was trying to set based on the current value, but I ended up having to do pretty much what the other workarounds said, while also having to flip the `bool` value back to false.

`context.ReadAsButtonValue()` is true during the start phase, but not when the performed event goes through. `context.action.triggered` is true for performed, though. The end result is the same, though. I'd prefer to see a lifecycle similar to:

1. event started
2. event continuing (possibly, but i'm not using it)
3. event performed
4. event complete.

Basically, if the values come through using events, it'd be nice to get an update notification when the button is done being pressed _separate from the performed event_ so i can change my bool back to false there instead of in the logic that uses that value.

**Comment by ian-rosenberg** [comment]
I noticed this issue in Unity 2019.3.0f6. My keyboard events all fire twice, indicated by both double console messages, and hitting the same breakpoint twice.

**Comment by LexGear** [comment]
This issue is driving me nuts. I've been at it for weeks. For me, I only have this issue with whichever controller is player one. Even if I use keyboard, if it's the first player in, it fires twice as often.

**Comment by ericcontainer** [comment]
> This issue is driving me nuts. I've been at it for weeks. For me, I only have this issue with whichever controller is player one. Even if I use keyboard, if it's the first player in, it fires twice as often.

`
public void TestInput(InputAction.CallbackContext ctx)
    {
        if(!ctx.performed)
            return;
	}
`
This solution from @OmegaRogue it worked like a charm. I'm using Input System 1.0.0. Unity 2020.2.2a18

**Comment by catsmu** [comment]
Likely unrelated to the solution @OmegaRogue presented, but I ran into the same issue when using the PlayerInputManager's PlayerInput prefab instantiation functionality, and noticed that the callback was firing on both the instanced PlayerInput component **and** the PlayerInput component on the prefab asset. Checking for `gameObject.scene.IsValid()` when processing the action callback appears to work around the issue, but a proper fix would certainly be preferred.

**Comment by LexGear** [comment]
> Likely unrelated to the solution @OmegaRogue presented, but I ran into the same issue when using the PlayerInputManager's PlayerInput prefab instantiation functionality, and noticed that the callback was firing on both the instanced PlayerInput component **and** the PlayerInput component on the prefab asset. Checking for `gameObject.scene.IsValid()` when processing the action callback appears to work around the issue, but a proper fix would certainly be preferred.

@catsmu You might be the first person in all the forms I've posted in to actually read my comment and not present the same "check for performed" solution. This sounds exactly like what could be happening. I'll try your fix and I'll also test on non-prefab objects and see if this is the issue.

**Comment by ps2goat** [comment]
I didn't use the normal prefab route, just newed up an instance of my input
maps in the start method of my player controller. This was how the
resources i used were doing it. That doesn't mean it's correct now since
all of those were during the alpha/beta.

But I'd like to think I had the simplest of examples and wanted to get it
working before messing around with multiple users or dynamic instantiation.
I'm sure it works somehow, and we need to write the manual 🌝

**Comment by LexGear** [comment]
> Likely unrelated to the solution @OmegaRogue presented, but I ran into the same issue when using the PlayerInputManager's PlayerInput prefab instantiation functionality, and noticed that the callback was firing on both the instanced PlayerInput component **and** the PlayerInput component on the prefab asset. Checking for `gameObject.scene.IsValid()` when processing the action callback appears to work around the issue, but a proper fix would certainly be preferred.

Catsmu, you are a legend! Thank you. Thank you. Thank you. Not even Unity's own staff could help me with this problem. They were completely stumped. Which is very, very odd, as you'd expect everyone would be using a prefab to inject a player. Thank you again.

**Comment by catsmu** [comment]
Cool, glad to hear it worked for you! Yeah, bit surprised this slipped through since it's the recommended pattern when using PlayerInputManager. Ah well. Are you aware of a thread on the Unity issue tracker that one of us can comment on? Or maybe @stefanunity can run with it from here?

**Comment by ps2goat** [comment]
I just got back to this, and apparently my problem was just some poorly phrased UI labels and weird layouts.

My point of using the new Input System was to be able to do things at a high level. So I have an Action of "Interact", and want to set what happens at that level, not at the binding level of each control type. 

At the Action level, all I _had_ (see below!) for actions were `Value`, `Button`, and `Pass Through`. I added a Unity `Interaction` to try and get what I wanted, and I originally used `Press and Release`.  I thought this meant the same thing as `Hold`, which I didn't have in my list. I wanted the player to press the key, and when it released, the action would happen.

-------------------------------------------
So now the fun, and even weirder, stuff. My original issue was around May 13. Today, August 31, it seems like I have more options. When I clicked on an individual binding under an Action, I was able to select an interaction type of `Tap`, and I see the other types. My previous options of `Release`, and `Press and Release` are gone!

When I went back up to the Action level to verify my options (because I feel like I'm crazy, now!), those were still the old ones with Press and Release, e.g. But then I saved and exited the Input System GUI to test, and it worked with just `Release`. Then I tried it again to look at my available Interactions, and now the Action level has the same options as the individual bindings!

It works with the `Tap` interaction, and it works without any interactions, now!

TLDR;
There's something funky with the UI. I did not explicitly download any new updates or Unity versions; I'm still on the same one unless there were background updates. Weird. Version 2019.3.13f1

The Input System package is still 1.0.0

**Comment by LexGear** [comment]
> I just got back to this, and apparently my problem was just some poorly phrased UI labels and weird layouts.
> 
> My point of using the new Input System was to be able to do things at a high level. So I have an Action of "Interact", and want to set what happens at that level, not at the binding level of each control type.
> 
> At the Action level, all I _had_ (see below!) for actions were `Value`, `Button`, and `Pass Through`. I added a Unity `Interaction` to try and get what I wanted, and I originally used `Press and Release`. I thought this meant the same thing as `Hold`, which I didn't have in my list. I wanted the player to press the key, and when it released, the action would happen.
> 
> So now the fun, and even weirder, stuff. My original issue was around May 13. Today, August 31, it seems like I have more options. When I clicked on an individual binding under an Action, I was able to select an interaction type of `Tap`, and I see the other types. My previous options of `Release`, and `Press and Release` are gone!
> 
> When I went back up to the Action level to verify my options (because I feel like I'm crazy, now!), those were still the old ones with Press and Release, e.g. But then I saved and exited the Input System GUI to test, and it worked with just `Release`. Then I tried it again to look at my available Interactions, and now the Action level has the same options as the individual bindings!
> 
> It works with the `Tap` interaction, and it works without any interactions, now!
> 
> TLDR;
> There's something funky with the UI. I did not explicitly download any new updates or Unity versions; I'm still on the same one unless there were background updates. Weird. Version 2019.3.13f1
> 
> The Input System package is still 1.0.0

Oof, I'd probably be setting up a new thread for this issue since this one is now closed and also because this is a whole different ball game.

Would be good to see your code. Because, and I'm not sure about this, but maybe you might be going about this the wrong way. 

I will say this... the new input system is buggy. It's not just you. It's buggy. I ask myself what even is the point of half these options when the program ignores it anyway.

**Comment by R1ckyCZ** [comment]
I've just encountered the same issue, was playing with it for a while and it seems like it got fixed for me when I changed action type from Button to Pass Through -> Button

Like this:
![image](https://user-images.githubusercontent.com/19607505/92704323-2b268000-f353-11ea-9b68-9176dfe349b5.png)

Maybe it will fix it also for others

**Comment by rafaelgonzalez** [comment]
> I've just encountered the same issue, was playing with it for a while and it seems like it got fixed for me when I changed action type from Button to Pass Through -> Button
> 
> Like this:
> ![image](https://user-images.githubusercontent.com/19607505/92704323-2b268000-f353-11ea-9b68-9176dfe349b5.png)
> 
> Maybe it will fix it also for others

Yes, that is what was triggering my action twice for me too!

Just like @OmegaRogue pointed, I would get an event twice, once with `context.started` being `true` and another with `context.performed` being `true`.

I originally used @OmegaRogue's solution, but rather than dealing with the double trigger in code, your solution avoids it altogether! The event is only triggered once and `context.started` is now always `false`. 👍 

---

@stefanunity I'd like to point out that this was a default setting as set by the input system when you first install it (in my case the "Move" action; in @Ricky-Brno's case above, the default "Fire" action)
